filament/backend/src/opengl/platforms/CocoaExternalImage.mm

```diff
@@ -0,0 +1,236 @@
+    gl_Position=vec4(x, y, 0.0, 1.0);
+}
+)SHADER";
+
+static const char *s_fragment = R"SHADER(#version 410 core
+precision mediump float;
+
+uniform sampler2DRect bgraSampler;
+layout(location = 0) out vec4 fragColor;
+
+void main() {
+    vec4 texColor = texture(bgraSampler, gl_FragCoord.xy);
+    fragColor = texColor.bgra;
+}
+)SHADER";
+
+CocoaExternalImage::SharedGl::SharedGl() noexcept {
+    glGenSamplers(1, &sampler);
+    glSamplerParameteri(sampler, GL_TEXTURE_MIN_FILTER, GL_NEAREST);
+    glSamplerParameteri(sampler, GL_TEXTURE_MAG_FILTER, GL_NEAREST);
+    glSamplerParameteri(sampler, GL_TEXTURE_WRAP_S, GL_CLAMP_TO_EDGE);
+    glSamplerParameteri(sampler, GL_TEXTURE_WRAP_T, GL_CLAMP_TO_EDGE);
+    glSamplerParameteri(sampler, GL_TEXTURE_WRAP_R, GL_CLAMP_TO_EDGE);
+
+    GLint status;
+
+    vertexShader = glCreateShader(GL_VERTEX_SHADER);
+    glShaderSource(vertexShader, 1, &s_vertex, nullptr);
+    glCompileShader(vertexShader);
+    glGetShaderiv(vertexShader, GL_COMPILE_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    fragmentShader = glCreateShader(GL_FRAGMENT_SHADER);
+    glShaderSource(fragmentShader, 1, &s_fragment, nullptr);
+    glCompileShader(fragmentShader);
+    glGetShaderiv(fragmentShader, GL_COMPILE_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    program = glCreateProgram();
+    glAttachShader(program, vertexShader);
+    glAttachShader(program, fragmentShader);
+    glLinkProgram(program);
+    glGetProgramiv(program, GL_LINK_STATUS, &status);
+    assert_invariant(status == GL_TRUE);
+
+    // Save current program state.
+    GLint currentProgram;
+    glGetIntegerv(GL_CURRENT_PROGRAM, &currentProgram);
+
+    glUseProgram(program);
+    GLint samplerLoc = glGetUniformLocation(program, "bgraSampler");
+    glUniform1i(samplerLoc, 0);
+
+    // Restore state.
+    glUseProgram(currentProgram);
+}
+
+CocoaExternalImage::SharedGl::~SharedGl() noexcept {
+    glDeleteSamplers(1, &sampler);
+    glDetachShader(program, vertexShader);
+    glDetachShader(program, fragmentShader);
+    glDeleteShader(vertexShader);
+    glDeleteShader(fragmentShader);
+    glDeleteProgram(program);
+}
+
+CocoaExternalImage::CocoaExternalImage(const CVOpenGLTextureCacheRef textureCache,
+        const SharedGl &sharedGl) noexcept : mSharedGl(sharedGl), mTextureCache(textureCache) {
+    glGenFramebuffers(1, &mFBO);
+    CHECK_GL_ERROR(utils::slog.e)
+}
+
+CocoaExternalImage::~CocoaExternalImage() noexcept {
+    glDeleteFramebuffers(1, &mFBO);
+    release();
+}
+
+bool CocoaExternalImage::set(CVPixelBufferRef image) noexcept {
+    // Release references to a previous external image, if we're holding any.
+    release();
+
+    if (!image) {
+        return false;
+    }
+
+    OSType formatType = CVPixelBufferGetPixelFormatType(image);
+    ASSERT_POSTCONDITION(formatType == kCVPixelFormatType_32BGRA,
+            "macOS external images must be in either 32BGRA format.");
```

Comment:
It looks like the ONLY format supported is 32BGRA, why does it say "either"?

Reply:
It's my mistake. I copy source code from CocoaTouchExternalImage.mm and modify for macOS. "either" is coming along from copy and paste.